### ISSUE TITLE
[LSR] Assertion failure: No dominance between recurrences used by one SCEV?

### ISSUE THREAD

**Issue by vzakhari** [llvm/llvm-project#60927]
The issue affects SPEC CPU2017/521.wrf compiled with Flang.  The assertion started kicking in after https://reviews.llvm.org/D144014

This is llvm-reduce'd reproducer: [reduced.ll.gz](https://github.com/llvm/llvm-project/files/10808264/reduced.ll.gz)

I was not able to reproduce the issue with an `opt` invocation, so here is the `clang` command to reproduce: `clang -cc1 -emit-obj -O2 -x ir reduced.ll`
```
clang: /llvm-project/llvm/lib/Analysis/ScalarEvolution.cpp:744: std::optional<int> CompareSCEVComplexity(llvm::EquivalenceClasses<const llvm::SCEV*>&, llvm::EquivalenceClasses<const llvm::Value*>&, const llvm::LoopInfo*, const llvm::SCEV*, const llvm::SCEV*, llvm::DominatorTree&, unsigned int): Assertion `DT.dominates(RHead, LHead) && "No dominance between recurrences used by one SCEV?"' failed.
```

If I revert 7abe3497e72af3ddee789dfc62c63a981a25dbf6 the compilation passes.

@davemgreen can you please take a look?

**Comment by vzakhari** [comment]
@kiranchandramohan, FYI.  This may affect `aarch64`, though, I have not checked this.

**Comment by kiranchandramohan** [comment]
@vzakhari I could not reproduce this on `aarch64` with today's build.

**Comment by davemgreen** [comment]
Thanks for the report. I will take a look. Form the IR it looks like there are multiple addrecs from different loops.

**Comment by vzakhari** [comment]
Thank you, David!

**Comment by kiranchandramohan** [comment]
I am probably missing a point here, but why is the `false` there in this expression?
```
      (false && isa<SCEVAddRecExpr>(Best) && isa<SCEVAddRecExpr>(Reg) &&
       cast<SCEVAddRecExpr>(Best)->getLoop() !=
           cast<SCEVAddRecExpr>(Reg)->getLoop())
```

**Comment by davemgreen** [comment]
Oh - no. That was a mistake I had for testing. That probably means my test isn't working properly either. Thank for catching it.

**Comment by davemgreen** [comment]
OK it is now hopefully properly fixed in 74b67e53c638fae6542372681cc181f1f4e23f12. Please let me know if it's not or anything else looks wrong. Thanks.

**Comment by vzakhari** [comment]
Thank you, David! The issue is gone.
